### PR TITLE
Fixes unsoundess issue in method overrides due to hierarchy of @PolyDet and @PolyDet("use")

### PR DESCRIPTION
**Pull request by t-rasmud**

### CODE REVIEW

File: docs/manual/figures/poly-hierarchy.svg [hunk 1]
```diff
@@ -0,0 +1,249 @@
+<?xml version="1.0" encoding="UTF-8" standalone="no"?>
```
Comment (smillst) [inline docs/manual/figures/poly-hierarchy.svg]: The file name should include determinism in the name.  Also the top two boxes are cut off. (https://github.com/t-rasmud/checker-framework/blob/bf4429bec883f6b50e2bef0373b21f37334d8cc3/docs/manual/figures/poly-hierarchy.svg)

File: docs/manual/determinism-checker.tex [hunk 2]
```diff
@@ -448,9 +448,9 @@
 myMethod(detList , index);
 \end{Verbatim}
 
-\subsectionAndLabel{Disallowing \<OrderNonDet> in polymorphic instantiations}{determinism-no-ordernondet}
+\subsectionAndLabel{Disallowing \<@OrderNonDet> in polymorphic instantiations}{determinism-no-ordernondet}
```
Comment (smillst) [inline docs/manual/determinism-checker.tex]: Could you please push the unrelated manual changes to the main branch so that they don't clutter the diffs here?

File: docs/manual/determinism-checker.tex [hunk 3]
```diff
@@ -469,14 +469,34 @@
...
+        \includeimage{poly-hierarchy}{6cm}
+    \end{center}
+    \caption{\<@PolyDet> type hierarchy}
+    \label{fig-poly-hierarchy}
+\end{figure}
+Figure~\ref{fig-poly-hierarchy} shows the type qualifier hierarchy among all the
+variants of \<@PolyDet>. Notice that \<@PolyDet> and \<@PolyDet("use")> are considered
+to be same in this hierarchy.
+Consider a method \code{m(@PolyDet ..., @PolyDet("use") ...)} that is overridden as
+\code{m(@PolyDet ..., @PolyDet("use") ...)}. Since \code{@PolyDet} is considered to be
```
Comment (smillst) [inline docs/manual/determinism-checker.tex]: The two methods have the same signature; one of them should be changed, but I don't know which.

File: docs/manual/determinism-checker.tex [hunk 4]
```diff
@@ -469,14 +469,34 @@
...
+the same as \code{@polyDet("use)} in the determinism type hierarchy, this overriding
+relationship is valid. But this is incorrect as it allows the following instantiations:
+Overridden method could be \code{m(@NonDet ..., @Det ...)} and the overrider method
```
Comment (smillst) [inline docs/manual/determinism-checker.tex]: This is confusing.  I think it would be clearer to give an example of the call that should be illegal.

File: docs/manual/determinism-checker.tex [hunk 5]
```diff
@@ -469,14 +469,34 @@
...
 \begin{Verbatim}
 @OrderNonDet List<@Det Integer> ondList;         // Suppose ondList contains elements: [1, 2, 3]
 ondList.remove(0);
 \end{Verbatim}
-Since the list is \<OrderNonDet>, removing an element at index 0 could remove any of its 3 elements in different runs
-making the resulting list \<NonDet> which is unsound.
+Since the list is \<@OrderNonDet>, removing an element at index 0 could remove any of its 3 elements in different runs
+making the resulting list \<@NonDet> which is unsound.
+
+\subsectionAndLabel{Type qualifier hierarchy among \<@PolyDet> variants}{determinism-poly-hierarchy}
```
Comment (smillst) [inline docs/manual/determinism-checker.tex]: This section name does not match it contents.  This section is about an additional overriding rule.

File: docs/manual/determinism-checker.tex [hunk 6]
```diff
@@ -469,14 +469,34 @@
...
+\begin{figure}
```
Comment (smillst) [inline docs/manual/determinism-checker.tex]: This figure should probably be moved to the section that introduces the polymorphic qualifiers.

File: docs/manual/determinism-checker.tex [hunk 7]
```diff
@@ -476,7 +488,34 @@
...
+relationship is valid. However, it allows the following unsound instantiations:
+\begin{Verbatim}
+    void caller(@NonDet String nStr, @Det String dStr) {
+        Super superObj = new Super();
+        Sub subObj = new Sub();
+        superObj.m(nStr, dStr);
+        subObj.m(dStr, nStr);
+    }
+\end{Verbatim}
+The first argument of \code{subObj.m} is not a supertype of the
```
Comment (smillst) [inline docs/manual/determinism-checker.tex]: This does not explain the unsoundness. Why does it matter that the first arg of subObj.m is not a supertype of the first argument of superObjt.m? 

Also, it's confusing that both methods have `@PolyDet("use")`.  Only the method in the sub class should have `@PolyDet("use")`

**Comment by t-rasmud** [comment]
Diverged from original code.
